mptcp: Earliest Completion First (ECF) Scheduler

dweb32 opened the pull request:
ECF was initially proposed and evaluated in this scientific publication:

> Lim, Y. S., Nahum, E. M., Towsley, D., & Gibbens, R. J. (2017, November). ECF: An MPTCP path scheduler to manage heterogeneous paths. In Proceedings of the 13th International Conference on emerging Networking EXperiments and Technologies (pp. 147-159). ACM.

This pull request is based on the implementation that was published by the original authors for MPTCP v0.89

I ported the code to MPTCP 0.94 and then to mptcp_trunk and and tried to align it as close as possible to the official MPTCP project.

### Note: That this scheduler is still very experimental!

dweb32 committed:
mptcp: Earliest Completion First (ECF) Scheduler

This scheduler works much like the default MPTCP scheduler. It always
prefers the subflow with the smallest round-trip-time that is available.

Signed-off-by: Daniel Weber <weberdaniel@gmx.net>

cpaasch commented:
First of all: Kudos to you, Daniel, for picking up these schedulers from the research papers and putting them in a form that can be included in the official code-repository! Keep it up! :)

(and now actually going to look at the code ;-) )

cpaasch commented on include/linux/tcp.h:
```diff
@@ -342,6 +342,7 @@ struct tcp_sock {
 	u32	snd_cwnd_clamp; /* Do not allow snd_cwnd to grow above this */
 	u32	snd_cwnd_used;
 	u32	snd_cwnd_stamp;
+	u32	snd_cwnd_before_idle_restart; /* used by MPTCP ECF scheduler */
```
This should probably go into ```mptcp_sched``` of ```struct mptcp_tcp_sock```.

dweb32 replied:
Sure, I will just put this in `ecfsched_priv`, where all the scheduling-related subflow information are located.

cpaasch commented on include/net/tcp.h:
```diff
@@ -1490,8 +1490,11 @@ static inline void tcp_slow_start_after_idle_check(struct sock *sk)
 	delta = tcp_jiffies32 - tp->lsndtime;
-	if (delta > inet_csk(sk)->icsk_rto)
+	if (delta > inet_csk(sk)->icsk_rto) {
+		tp->snd_cwnd_before_idle_restart =
```
We should try to find a cleaner way to update this variable without having to introduce changes to the regular TCP stack.

cpaasch replied:
Looking through the code, ```snd_cwnd_before_idle_restart``` is only used when ```ecf_optim``` is set. That latter one defaults to 0 however. Without having read the paper yet, I'm not sure if we need it.

dweb32 replied:
> Looking through the code, `snd_cwnd_before_idle_restart` is only used when `ecf_optim` is set. That latter one defaults to 0 however. Without having read the paper yet, I'm not sure if we need it.

I have read the paper again and `ecf_optim` is not mentioned as part of the scheduling algorithm. I conclude that this could only be a remainder of the author's evaluation to prove the effect of resetting the CWND by simply turning off their proposed behavior. Thus, we can remove this module parameter completely.

dweb32 replied:
> We should try to find a cleaner way to update this variable without having to introduce changes to the regular TCP stack.

@cpaasch I totally agree with you. Do you have any suggestions for properly identifying this event from a scheduling perspective?

cpaasch replied:
From what I understand, if you remove the ```ecf_optim```, there isn't a need anymore for this code here in ```include/net/tcp.h```.

dweb32 replied:
No, we still need to save the cwnd value. I have now moved it to ecfsched_priv. This is actually the main idea of the proposed algorithm to avoid slow-start (in the estimation) again.

Meanwhile I did some major cleanup of the oringinal code to make it more readable. The only thing i still did not solve is how to detect when the subflows cwnd was reset due to a idle-phase.

cpaasch replied:
Why do we need to save it? AFAICS, it is only read from when ```ecf_optim``` is set, no?

dweb32 replied:
I suggest to completely remove the `ecf_optim` module parameter and by default always use `max(besttp->snd_cwnd, ecf_p->snd_cwnd_before_idle_restart)` as it is proposed in the paper.

We need to save the `snd_cwnd` value because TCP resets it when the subflow is idling for a RTO. In that case, we estimate the completion time with the previously recorded value (in the paper called as `rec_snd_cwnd` or called in the initial implementation as `snd_cwnd_before_idle_restart`) in order not to underestimate the subflows performance.

cpaasch replied:
Ok, I got confused because ```ecf_optim``` was initialized to 0. So, the intention is to default-enable ```ecf_optim``` ?

If a subflow is idle such that the snd_cwnd gets reset, we don't know anymore what the capacity of that path is.
In that case, can we really assume that it is going to be the same capacity as before the idle-period? If the assumption in the network is that the capacity remains the same over idle periods, then one could simply disable restart-after-idle globally, no?
Further, during these idle-periods the congestion-window is only decreasing gradually by a factor of 2 per RTO.

Looking at the paper [1], I don't see a reference to the restart-after-idle part. Maybe you can start an e-mail thread with the authors (CC me and Matt) to clarify this part ?

I'm hesitant to introduce this kind of change to the TCP-stack without being convinced that it is really necessary.


[1] https://www.repository.cam.ac.uk/bitstream/handle/1810/279113/ecf_extended_abstract_rev1.pdf?sequence=1

dweb32 replied:
> Looking at the paper [1], I don't see a reference to the restart-after-idle part.

@cpaasch I just noticed that we are looking at two different versions of the same paper. Your are looking at the one that was published at SIGMETRICS’17 [1] and I am looking at the longer one published at CoNEXT ’17 [2]. The later describes the effect of idle subflows in detail in section 3.2

> Ok, I got confused because ecf_optim was initialized to 0. So, the intention is to default-enable ecf_optim ?

Yes. At least this is the way I understand the conclusion given in [2] and therefore would suggest to remove it completely. Because when you choose to use ECF instead of BLEST or the default scheduler, this is the feature you actually expect to use, or don't you?

> If a subflow is idle such that the snd_cwnd gets reset, we don't know anymore what the capacity of that path is.
In that case, can we really assume that it is going to be the same capacity as before the idle-period? If the assumption in the network is that the capacity remains the same over idle periods, then one could simply disable restart-after-idle globally, no?

Good point. If we disable restart-after-idle globally (in the case of almost constants capacities) we may prevent the performance degrade that was observed in [2] and achieve similar results. Thus, we could just remove `ecf_optim` and `snd_cwnd_before_idle_restart`.

> I'm hesitant to introduce this kind of change to the TCP-stack without being convinced that it is really necessary.

I can absolutely understand your view and I also do not want to modify the TCP-stack at all. I was strongly encouraged by @teto and @matbee to share the ECF version that I (somehow) ported to a current version so far. Unfortunately, I got my hands on the original ECF implementation so late, that it didn't made it into my evaluation.

I will update this pull-request with a more streamlined version of ECF matching the other schedulers. Until now I sticked very close to the orginal code published for v0.89

[1] https://dl.acm.org/citation.cfm?doid=3078505.3078552
[2] https://dl.acm.org/citation.cfm?doid=3143361.3143376

cpaasch reviewed (changes requested):
First pass complete - awaiting some feedback on my after-idle questions.

dweb32 commented:
I just updated the PR and rebased it on top of the current mptcp_trunk branch. @cpaasch, can you have a look again and review this version please?

I also added the original code s6dlwebe@6c1f5d1493afb5921c1151e20417f18db78c435f because it is no longer available from the authors webpage given in the paper.

matttbe commented on net/mptcp/mptcp_ecf.c:
```diff
@@ -0,0 +1,384 @@
+
+static unsigned int r_beta __read_mostly = 4; /* beta = 1/r_beta = 0.25 */
+module_param(r_beta, int, 0644);
+MODULE_PARM_DESC(r_beta, "beta for ECF");
```
may you prefix the name of the module please?

    mptcp_ecf_r_beta

The main reason is that if the file is not compiled as a module but included in the kernel image, the only possibility to change this parameter is to add a parameter when you boot the kernel. Then you want to avoid having "generic" names that could conflict with others.

dweb32 replied:
Ok, I will change this. But the same also applies to the module parameters of the BLEST an RR scheduler as well.

matttbe replied:
Good point. Do not hesitate to fix that for the other schedulers.

@cpaasch : or do you think it is not needed?

cpaasch replied:
@matttbe - I think that would be good to change.
@s6dlwebe - Let us know if you want to do that change in a separate PR. Otherwise, I will add it to the generic todo-list.

dweb32 replied:
> @matttbe - I think that would be good to change.
> @s6dlwebe - Let us know if you want to do that change in a separate PR. Otherwise, I will add it to the generic todo-list.

I just updated the PR onto mptcp_trunk and added the `mptcp_ecf_` prefix to the module parameter. For now, all schedulers are written in the same way. As @matttbe already pointed out, all schedulers use a mix of names for internal structures and static functions that we should change.

@cpaasch - I would prefer a separate PR for cleaning up all the scheduler naming mixup and add a todo to reduce the code duplication later. This will make it easier to keep track of the name changes and refactoring.

matttbe replied:
Sounds good to me!

matttbe replied:
and thank you for the update! :)

matttbe commented on net/mptcp/mptcp_ecf.c:
```diff
@@ -0,0 +1,384 @@
+module_param(r_beta, int, 0644);
+MODULE_PARM_DESC(r_beta, "beta for ECF");
+
+struct ecfsched_priv {
```
detail: there is a mix of names for the "internal" structures and static functions. Maybe better to uniform that if you don't mind, e.g. having everything starting with `mptcp_ecf_` (or something else)

dweb32 replied:
Ok, I had already done a unification once in fa1d0ee656d7558bf032ef3924a003681b661f17 where I applied the style of `mptcp_sched.c` to all other schedulers. I can apply a `mptcp_` + [scheduler] prefix to all to unify them.

matttbe replied:
I think if you can be at least consistent with the existing code, that's good.
Having a good prefix is useful when you read call traces or if you want to do some function tracing.

matttbe commented on net/mptcp/mptcp_ecf.c:
```diff
@@ -0,0 +1,384 @@
+				ecf_p->last_rbuf_opti = tcp_jiffies32;
+			}
+		}
+	}
```
I think a part of this function is also used in the default scheduler and in BLEST. Do you think we could share some code here?

dweb32 replied:
Yes, BLEST and ECF share a lot of code with the default scheduler. The problem is that every scheduler (re-)defines its own priv where `last_rbuf_opti` is stored.

But although reusing the `defsched_priv` of the default scheduler should work here, it is not a very clean solution. So I do not have a better idea right now without introducing bigger changes to `mptcp_sched.c`.

matttbe replied:
if the only "issue" is the private structure, you can:
- give a pointer to the shared function: `*last_rbuf_opti = tcp_jiffies32;`
- have a "common" structure that schedulers can expend (like what it is usually done in the network stack: the socket structure is included in another one, included in TCP sock, included in MPTCP one.

dweb32 replied:
@matttbe have a look now, meanwhile I removed many redundant code blocks...

for now it works, but I think we should change it to an expendable common structure

matttbe commented on net/mptcp/mptcp_ecf.c:
```diff
@@ -0,0 +1,384 @@
+	return NULL;
+}
+
+/* copy from mptcp_sched.c: __mptcp_next_segment */
```
if it is a copy, can we not share code?

EDIT: after a quick look, it seems it is not a copy. There should be a pointer to which `get_available_subflow()` function to use if we want to share code.

dweb32 replied:
Well, when I worked on BLEST, @cpaasch was fine with some code-duplication https://github.com/multipath-tcp/mptcp/pull/300#issuecomment-444616756

matttbe replied:
I was fine as well. I can still be fine here but I prefer to ask the question because that's the second time we could do that :-)

@cpaasch: what do you think? :-)

dweb32 replied:
@matttbe removed this block completely, since `mptcp_next_segment` is now called directly via `sched_ops`

matttbe commented on net/mptcp/mptcp_ecf.c:
```diff
@@ -0,0 +1,384 @@
+	return skb;
+}
+
+/* copy from mptcp_sched.c: mptcp_next_segment */
```
same here, could we share code to avoid copy? (I didn't check how much was copied)

dweb32 replied:
the same as above... only three lines changed here, as we currently have no possibility to properly hook `ecf_get_available_subflow` into the default scheduler and therefore must copy the rest as well.

matttbe replied:
you can also give a pointer to the function you want to execute as a parameter of the shared function, no?

If there are many differences, it is also possible to split the function in smaller bits.

If it looks similar but many small differences, yes, better to copy maybe.

matttbe replied:
What is clear is that the scheduler code is not easy to understand. Better to avoid too much abstraction. But if it can reduce a lot the code of other schedulers to only focus on what's particular to this scheduler, that's could also be good :-)

matttbe reviewed (commented):
That looks good! Thank you for working on that!
Also good we no longer have modified code in TCP stack.

Here are a few ideas to uniform the code and maybe reduce the amount of code in your new mptcp_ecf.c code.

cpaasch commented:
As for the code-duplication question: I think it is fine for now. I would suggest we merge ECF and then take a step back and see what really are "generic" scheduling decisions that we can factor out.

matttbe commented:
> As for the code-duplication question: I think it is fine for now. I would suggest we merge ECF and then take a step back and see what really are "generic" scheduling decisions that we can factor out.

Sounds good to me!

dweb32 committed:
mptcp: Reduce code-duplication for other schedulers

'mptcp_next_segment' now honors the function pointer to the actual part
that makes the scheduling decision in 'sched_ops->get_subflow'. This
allows for a better reuse by other schedulers.

The BLEST scheduler needs to adapt the direction of lambda value change
depending on the occurrence of a retransmission. In order to remove the
copied 'mptcp_rcv_buf_optimization' as well the scheduler now checks the
tcp 'retrans_stamp' of the meta socket.

Signed-off-by: Daniel Weber <weberdaniel@gmx.net>

dweb32 committed:
mptcp: Remove code-duplication from ECF scheduler

The ECF scheduler relies on large parts of the default scheduler. This
commit removes the copied blocks and reuses 'mptcp_next_segment' and
'mptcp_rcv_buf_optimization' directly from it via function pointers.

Signed-off-by: Daniel Weber <weberdaniel@gmx.net>

cpaasch commented:
Sorry for the way too long delay... I applied this to mptcp_trunk. Changes outside of ecf and blest look very fine to me. If there are any remaining issues with ECF we can fix them in mptcp_trunk.
